### PR TITLE
Add MP3 support to SDL2_mixer

### PR DESCRIPTION
**Pull request by pudinha**

### CODE REVIEW

File: tools/ports/mpg123.py [hunk 1]
```diff
@@ -0,0 +1,2337 @@
+# Copyright 2016 The Emscripten Authors.  All rights reserved.
```
Comment (kripken) [inline tools/ports/mpg123.py:1]: this should be `2020`

File: tools/ports/sdl2_mixer.py [hunk 2]
```diff
@@ -30,11 +30,28 @@ def create():
     shutil.copytree(source_path, dest_path)
 
     final = os.path.join(dest_path, libname)
-    ports.build_port(dest_path, final, [], ['-DOGG_MUSIC', '-O2', '-s', 'USE_VORBIS=1', '-s', 'USE_SDL=2'],
-                     ['dynamic_flac', 'dynamic_fluidsynth', 'dynamic_mod', 'dynamic_modplug', 'dynamic_mp3',
-                      'fluidsynth', 'load_mp3', 'music_cmd', 'music_flac', 'music_mad', 'music_mod',
-                      'music_modplug', 'playmus.c', 'playwave.c'],
```
Comment (kripken) [inline tools/ports/sdl2_mixer.py:36]: what is the impact of removing `dynamic_modplug` and the other stuff removed here?

File: tools/ports/sdl2_mixer.py [hunk 3]
```diff
@@ -30,11 +30,28 @@ def create():
...
-                     ['external', 'native_midi', 'timidity'])
+    ports.build_port(
+      dest_path,
+      final,
+      includes=[],
+      flags=[
+        '-s', 'USE_SDL=2',
+        '-s', 'USE_MPG123=1',
```
Comment (sbc100) [inline tools/ports/sdl2_mixer.py:39]: We probably don't want to always build mp3 support here.  Maybe we can make two version of the mixer library?

**Comment by sbc100** [comment]
I was kind of hoping to avoid adding more ports until we can have optional/contrib ports

These would be ports that:

1. we don't test are part of normal CI (they could have their own tests)
2. Are not built as part of  `./embuilder build ALL` 
3. Could maybe life out-of-tree.

This change has re-invigorating my interested in getting this done.    However, since its not done yet I'm not sure how much sway I have to block this landing first :)

File: tools/ports/sdl2_mixer.py (hunk 2 above)
Comment (pudinha) [reply tools/ports/sdl2_mixer.py:36]: Those files are wrapped in
```
#ifdef MUSIC_MOD_MODPLUG
```
and similars, so they add nothing unless the macros are defined.

Note that most of those files were renamed ~ `dynamic_modplug` was renamed to `music_modplug`

File: tools/ports/sdl2_mixer.py (hunk 3 above)
Comment (pudinha) [reply tools/ports/sdl2_mixer.py:39]: I could follow the pattern on SDL2_IMAGE_FORMATS:
```
// Formats to support in SDL2_mixer. Valid values: ogg, mp3
var SDL2_MIXER_FORMATS = [];
```
Should I put *ogg* as a default, to maintain compatibility?
What do you think?

Comment (sbc100) [reply tools/ports/sdl2_mixer.py:39]: I think thats a good idea yes.  Most people don't want more than once audio format for a given project I imagine.

**Comment by pudinha** [comment]
> I was kind of hoping to avoid adding more ports until we can have optional/contrib ports
> 
> These would be ports that:
> 
> 1. we don't test are part of normal CI (they could have their own tests)
> 2. Are not built as part of  `./embuilder build ALL`
> 3. Could maybe live out-of-tree.
>
> This change has re-invigorating my interested in getting this done. However, since its not done yet I'm not sure how much sway I have to block this landing first :)

That sounds good to me. Having stuff out-of-tree could make it less intimidating for people to contribute.

Just one question though: why don't the ports use `emconfigure ./configure` and `emmake make install`? I was expecting them to be just a bunch of shell scripts like that ~ works for `SDL2_mixer` and `mpg123`.

:sparkles: :sparkles: :sparkles:

**Comment by sbc100** [comment]
Because we need to support windows and windows can't run autoconf and friends.  (at least not without a lot of not-so-common software such as cygwin).

File: tools/ports/sdl2_mixer.py (hunk 3 above)
Comment (pudinha) [reply tools/ports/sdl2_mixer.py]: done! :sparkles:

File: tools/ports/mpg123.py (hunk 1 above)
Comment (pudinha) [reply tools/ports/mpg123.py]: done! :sparkles:

**Comment by pudinha** [comment]
> I don't hear audio in the tests, though - maybe the sound is short, and by the time I click on the page (which should enable audio) they are over? This is on chrome 83.

@kripken, I tried chromium here and had problems, but with all formats (ogg and wav), not only mp3.

For me it had to do with the chromium expecting user interaction before allowing autoplay of sound ~ there are some exceptions on js console ~, and sometimes it's glitchy. I managed to get it running by quickly clicking around the page before the test loaded. Firefox works without intervention.

**Comment by kripken** [comment]
@pudinha strange, yeah, I also see firefox work but not chrome. The auto-resume feature might not work in SDL2-mixer then, but yeah, that's a separate issue.

File: tools/ports/sdl2_mixer.py [hunk 4]
```diff
@@ -52,7 +86,10 @@ def clear(ports, settings, shared):
 
 def process_dependencies(settings):
   settings.USE_SDL = 2
-  settings.USE_VORBIS = 1
+  if "ogg" in settings.SDL2_MIXER_FORMATS:
+    settings.USE_VORBIS = 1
+  if "mp3" in settings.SDL2_MIXER_FORMATS:
+    settings.USE_MPG123 = 1
```
Comment (sbc100) [inline tools/ports/sdl2_mixer.py:92]: You can conditionally add to deps here (see `tools/ports/sdl2_image.py`)

File: tests/test_interactive.py [hunk 5]
```diff
@@ -118,6 +118,13 @@ def test_sdl2_mixer_wav(self):
     Popen([EMCC, '-O2', '--minify', '0', os.path.join(self.get_dir(), 'sdl2_mixer_wav.c'), '-s', 'USE_SDL=2', '-s', 'USE_SDL_MIXER=2', '-s', 'INITIAL_MEMORY=33554432', '--preload-file', 'sound.wav', '-o', 'page.html']).communicate()
     self.run_browser('page.html', '', '/report_result?1')
 
+  def test_sdl2_mixer_mp3(self):
+    shutil.copyfile(path_from_root('tests', 'sounds', 'pudinha.mp3'), os.path.join(self.get_dir(), 'sound.mp3'))
+    open(os.path.join(self.get_dir(), 'sdl2_mixer_mp3.c'), 'w').write(self.with_report_result(open(path_from_root('tests', 'sdl2_mixer_mp3.c')).read()))
+
+    Popen([EMCC, '-O2', '--minify', '0', os.path.join(self.get_dir(), 'sdl2_mixer_mp3.c'), '-s', 'USE_SDL=2', '-s', 'USE_SDL_MIXER=2', '-s', 'SDL2_MIXER_FORMATS=["mp3"]', '-s', 'INITIAL_MEMORY=33554432', '--preload-file', 'sound.mp3', '-o', 'page.html']).communicate()
```
[the remark this answers is not in the document]
Comment (pudinha) [reply tests/test_interactive.py]: :sparkles:

File: tools/ports/sdl2_mixer.py (hunk 4 above)
Comment (pudinha) [reply tools/ports/sdl2_mixer.py:94]: :sparkles:

**Comment by pudinha** [comment]
@sbc100 I've rebased on top of master and adequated **embuilder.py** with 8fdee8e0e96df6317c2a2dc206a61df8f377345b. Could you please take a look? :sparkles:

**Pull request merged**